### PR TITLE
Advanced String format with named/numbered parameters

### LINKED ISSUE #5493: Advanced String format with named/numbered parameters
Someone asked for this feature on IRC is it something we could add to "String" or is too special?

```
func _ready():
    print(format("hello %0, what can %1 do for %0", {"0":"Unknown", "1":"NPC"}, "%_"))

func format(_str, dict, placeholder="[_]"):
    for key in dict:
        _str = _str.replacen( placeholder.replace("_", str(key) ), dict[key]) 
    return _str
```

### PR DESCRIPTION
**Pull request by AlexHolly**
closes #5493

There was some new stuff to me so maybe there is better way to solve this please double check.

**UPDATE** 
Tests

```
extends Node2D

func _ready():
	test01()

var t1 = "Hello {name},\nI am {magic_teacher1} your first teacher.\n{name} can get the title {title1}"
var t1_rs = "Hello RobotGuy,\nI am Logis your first teacher.\nRobotGuy can get the title rookie"

var t2 = "Hello {0},\nI am {1} your first teacher.\n{0} can get the title {2}."
var t2_rs = "Hello RobotGuy,\nI am Logis your first teacher.\nRobotGuy can get the title rookie."

var t3 = "Hello {0},\nI am {magic_teacher1} your first teacher.\n{0} can get the title {2}."
var t3_rs = "Hello RobotGuy,\nI am Logis your first teacher.\nRobotGuy can get the title 9000.91."

var t4 = "I know your name, [name] \nI am [magic_teacher1] your first teacher.\n[name] can get the title [title1]"
var t4_rs = "I know your name, RobotGuy \nI am Logis your first teacher.\nRobotGuy can get the title rookie"

var t5 = "I know your name, [0] \nI am [1] your first teacher.\n[0] can get the title [2]"
var t5_rs = "I know your name, RobotGuy \nI am Logis your first teacher.\nRobotGuy can get the title rookie"

var t6 = "Hello {name},\nI am {0} your first teacher.\n{name} can get the title {title1}"
var t6_rs = "Hello RobotGuy,\nI am Logis your first teacher.\nRobotGuy can get the title 9000.91"

var t7 = "Hello {0},\nI am {1} your first teacher.\n{0} can get the title {2}."
var t7_rs = "Hello RobotGuy,\nI am 42 your first teacher.\nRobotGuy can get the title 9000.91."
 
var t8 = "Hello 0[. I am 1[."
var t8_rs = "Hello RobotGuy. I am Logis."

var t9 = "Hello [0. I am [1."
var t9_rs = "Hello RobotGuy. I am Logis."

var t10 = "Hello {0}. I am {9000.91}."
var t10_rs = "Hello RobotGuy. I am Logis."

var t11 = "Hello {name}. I am {magic_teacher1}."
var t11_rs = "Hello RobotGuy. I am Logis."

var t12 = "Hello name[. I am magic_teacher1[."
var t12_rs = "Hello RobotGuy. I am Logis."

var t13 = "Hello [name. I am [magic_teacher1."
var t13_rs = "Hello RobotGuy. I am Logis."

var t14 = "Hello {name}. I am {magic_teacher1}, strength {strength}."
var t14_rs = "Hello RobotGuy. I am Logis, strength 9000.91."

func test01():
	#Test01 array with dict style
	assert(t1.format([["name","RobotGuy"],["magic_teacher1","Logis"],["title1","rookie"]]) == t1_rs)

	#Test02 with array
	assert(t2.format(["RobotGuy","Logis","rookie"]) == t2_rs)
	
	#Test03 mixed array
	assert(t3.format(["RobotGuy",["magic_teacher1","Logis"],9000.91]) == t3_rs)

	#Test04 custom placeholder with dict style
	assert(t4.format([["name","RobotGuy"],["magic_teacher1","Logis"],["title1","rookie"]],"[_]") == t4_rs)
	
	#Test05 custom placeholder with array style
	assert(t5.format(["RobotGuy","Logis","rookie"],"[_]") == t5_rs)
	
	#Test06 with string,int,real as key and value
	assert(t6.format([["name","RobotGuy"],[0,"Logis"],["title1",9000.91]]) == t6_rs)

	#Test07 with string,int,real as value
	assert(t7.format(["RobotGuy",42,9000.91]) == t7_rs)

	#Test08 postfix
	assert(t8.format(["RobotGuy","Logis"],"_[") == t8_rs)
	
	#Test09 prefix
	assert(t9.format(["RobotGuy","Logis"],"[_") == t9_rs)
	
	#Test10 with dictionary
	assert(t10.format({"0":"RobotGuy",9000.91:"Logis"}) == t10_rs)
	assert(t11.format({"name":"RobotGuy","magic_teacher1":"Logis"}) == t11_rs)
	assert(t12.format({"name":"RobotGuy","magic_teacher1":"Logis"},"_[") == t12_rs)
	assert(t13.format({"name":"RobotGuy","magic_teacher1":"Logis"},"[_") == t13_rs)
	assert(t14.format({"name":"RobotGuy","magic_teacher1":"Logis","strength":9000.91}) == t14_rs)
	
	#Test error messages
	#print("1234".format(["a",[0],"b",[0,1,2]]))
	#print("1234".format(Vector2(0,0)))

```

I'm thinking about adding Dictionary.to_array() to support more use cases.

Any thoughts on the whole thing?

### CODE REVIEW

**Comment by vnen** [comment]
I don't like this syntax. Why use an array of arrays instead of a dictionary? With a dictionary it will be simpler to read, simpler to parse, and no need to add a strangely named `Dictionary.to_array()` (which I can't tell what it does by the name).

Also, I would prefer the Python syntax of `{braces}` instead of `%percent`, as suggested in https://github.com/godotengine/godot/issues/5493#issuecomment-229464145.

**Comment by AlexHolly** [comment]
**Why array instead of a dictionary?**

One reason was to keep dictionary out of ustring.
The second reason was a functionality that makes this syntax possible.
```
#test with array
    print("Hello %0,\nI am %1 your first teacher.\n%0 can get the title %2.".format(["RobotGuy","Logis","rookie"],"%_"))
```

with an dictionary you would have to use {"0":"RobotGuy","1":"Logis","3":"rookie"} this is more work than ["RobotGuy","Logis","rookie"]


**What is Dictionary.to_array() for?**

Once "Dictionary.to_array()" is implemented you can use dictionarys like this

```
{"name":"RobotGuy","magic_teacher1":"Logis","title1":"rookie"}.to_array() -> [["name","RobotGuy"],["magic_teacher1","Logis"],["title1","rookie"]]
```
This is probably not the fastes way but I think it is the easiest for the users.


**Why %_ and not {_} as default?**

I have no opinion for the syntax there is only one reason I chose %_ godot already has a similar functionality that is using % http://docs.godotengine.org/en/latest/reference/gdscript_printf.html

**Comment by AlexHolly** [comment]
**Default syntax**

Vote this up for {_} 
Vote this down for %_

**Comment by bojidar-bg** [comment]
@AlexHolly You are basically granted to receive no downvotes :stuck_out_tongue:

**Comment by vnen** [comment]
> One reason was to keep dictionary out of ustring.

Well, since `dictonary.h` includes `ustring.h`, I guess it can't be as simple. But there might have a way around.

> with an dictionary you would have to use {"0":"RobotGuy","1":"Logis","3":"rookie"} this is more work than ["RobotGuy","Logis","rookie"]

The proposal was to use both with Dictionary and with Array, so in this case you could just pass an array. The function would adapt to each case.

**Comment by AlexHolly** [comment]
I thought I will get some typesafe error message from the editor when I use an Array instead of Variant (without running the game) but it looks like this is not happening so I can use this right?

```
String format(Variant values,String placeholder="{_}") const;
```

**Comment by AlexHolly** [comment]
So VARIANT is not working in "variant_call.cpp" 

"VARIANT is not defined"

and ```#include "dictionary.h"```  is not working in ustring.h

```
In file included from core/ustring.h:35:0,
                 from core/math/vector3.h:35,
                 from core/math/aabb.h:34,
                 from core/variant.h:36,
                 from core/object.h:33,
                 from core/os/input.h:32,
                 from platform/x11/os_x11.h:33,
                 from platform/x11/godot_x11.cpp:30:
core/dictionary.h:66:25: error: 'String' does not name a type
  Error parse_json(const String& p_json);
                         ^
core/dictionary.h:66:33: error: ISO C++ forbids declaration of 'p_json' with no type [-fpermissive]
  Error parse_json(const String& p_json);
                                 ^
core/dictionary.h:67:2: error: 'String' does not name a type
  String to_json() const;

```

Does anyone know how to solve this?

**Comment by Dillybob92** [comment]
Having a similar issue here with massively dynamic tooltips

I propose this type of syntax:
 
```
var format_string = "%customname1 was reluctant to learn %customname2, but now he enjoys it."
var actual_string = format_string % { 
customname1 = "Estragon",
 customname2 = "GDScript"
}
```

Real life example:
```
	GF.GTipRTL.set_bbcode("""[b]Fireball[/b][center][color=#bbbbbb]%s[/color][/center]
[b]Cooldown[/b]: [color=#e6e7d7]%ss[/color]
[b]Damage[/b]: [color=#e10707]%s[/color]
	""" % [C_SKILL.skill_description,
		  C_SKILL.skill_cooldown,
			""+str(C_SKILL.skill_damage[0])+" - "+str(C_SKILL.skill_damage[1])+""])
```

If I were to add another placeholder randomly I would need to **re-order my array** to match the order of the placeholders (`%s's`). Would be a PITA, but, I mean it's definitely fine now but maybe in the future `GDScript` can offer a dictionary type placeholder conversion @reduz ? :)

**Comment by bojidar-bg** [comment]
@AlexHolly Well, I guess, you can use advice from here: http://www.cplusplus.com/forum/articles/10627/#msg49680
More specifically:
```cpp
//a.h
#ifndef _A
#define _A
class A; // Forward-define so that we can circularly depend on it
#include "b.h"
class A {
    B* something;
}
#endif
```
```cpp
//b.h
#ifndef _B
#define _B
class B; // Forward-define so that we can circularly depend on it
// doing this on both sides since we don't know which would load first.
#include "a.h"
class B {
    A* something;
}
#endif
```

File: core/ustring.cpp [hunk 1]
```diff
@@ -2901,6 +2901,64 @@ bool String::matchn(const String& p_wildcard) const {
 
 }
 
+String String::format(const Variant& values,String placeholder) const {
+
+	String new_string = String( this->ptr() );
+
+	if( values.get_type() == Variant::ARRAY ) {
```
Comment (bojidar-bg) [inline core/ustring.cpp]: Shouldn't it be `->get_type` here?

File: core/variant_call.cpp [hunk 2]
```diff
@@ -1388,6 +1389,7 @@ _VariantCall::addfunc(Variant::m_vtype,Variant::m_ret,_SCS(#m_method),VCALL(m_cl
 	ADDFUNC0(STRING,STRING_ARRAY,String,bigrams,varray());
 	ADDFUNC1(STRING,REAL,String,similarity,STRING,"text",varray());
 
+  ADDFUNC2(STRING,STRING,String,format,NIL,"values",STRING,"placeholder",varray("{_}"));
```
Comment (bojidar-bg) [inline core/variant_call.cpp]: Indent is wrong here.

**Comment by bojidar-bg** [comment]
BTW, you might want to rebase the whole bunch of commits (+ forcepush), since Akien would ~~kill you~~ dislike it otherwise...

**Comment by akien-mga** [comment]
Superseded by #7010.

**Pull request closed without merging**